rama-http/src/layer/auth/async_require_authorization.rs

```diff
@@ -130,12 +130,28 @@ use std::future::Future;
 #[derive(Debug, Clone)]
 pub struct AsyncRequireAuthorizationLayer<T> {
     auth: T,
+    allow_anonymous: bool,
```

Comment:
This is the wrong layer/service to implement it for. As this is a generic layer which allows any kind of custom authorization. Good one if you want to implement your own authorization, but not in need of `allow_anonymous` as that would be something the used `auth: T` will implement themselves.

The proxy auth is good, that one you can keep.

You can still add it to `ValidateRequestHeaderLayer` and its service however. That one is about the `Authorization` header. Logic to be done there is pretty much what you did for `proxy_auth.rs`.

Reply:
Thank you very much!

Question on this [here](https://github.com/plabayo/rama/blob/main/rama-http/src/layer/validate_request/validate_request_header.rs#L176-L178) just to make sure I understand the intention: 

```rust
    async fn serve(
        &self,
        ctx: Context<State>,
        req: Request<ReqBody>,
    ) -> Result<Self::Response, Self::Error> {
        match self.validate.validate(ctx, req).await {
            Ok((ctx, req)) => self.inner.serve(ctx, req).await,
            Err(res) => Ok(res),
        }
    }
}
```
Is the expectation to insert `UserId::Anonymous` into the `Context` as part of the`Err(res)` arm? If so it seems like that differs from the logic of `proxy_auth.rs`  you had outlined in https://github.com/plabayo/rama/issues/330#issuecomment-2452376257

> If and only if the header used by that layer is absent and the bool is true we will authorize the user as UserID::Anonymous

In the `ValidateRequestHeader` service, if I'm understanding it correctly, there could be a situation where `validate(ctx, req)` fails because an `Accept: application/json` header isn't present. Would it follow that we would then still continue the chain of layers if `allow_anonymous == true` in that situation?

Something like this:

```rust
async fn serve(
        &self,
        ctx: Context<State>,
        req: Request<ReqBody>,
    ) -> Result<Self::Response, Self::Error> {
        match self.validate.validate(ctx, req).await {
            Ok((ctx, req)) => self.inner.serve(ctx, req).await,
            Err(res) => {
                if self.allow_anonymous {
                   // these are borrow after moves
                    ctx.insert(UserId::Anonymous);

                    self.inner.serve(ctx, req).await
                } else {
                    Ok(res)
                }
            }
        }
    }
```